### PR TITLE
Enable shared memory reuse in matmul epilogue

### PR DESCRIPTION
**Pull request by jacobhinkle**
This uses `promoteReuse` from #739 and inserts a syncthreads just before the epilogue loop when smem is used for the epilogue, when possible. The matmul heuristic attempts to predict when this will be possible in order to more accurately estimate shared memory usage, and hence occupancy. If we cannot guarantee re-use, we must assume in the heuristic that memory will not be reclaimed, even though it might be when the fusion is lowered.

Shared memory reclamation can only occur if the smem buffers have non-overlapping lifetimes. This is difficult to guarantee before scheduling and lowering. We use `cacheAfter` to create the `a` and `b` smem tiles, but we use `cacheBefore` for the epilogue smem tile. This means that smem will be  used for any downstream uses of `a` and `b` but the epilogue smem will have its lifetime restricted to the epilogue itself, regardless of downstream uses of the matrix product.

The uses of `a` and `b` can complicate lifetime analysis. Consider a case where both matrices are square and we wish to compute `a @ b + a` where `@` denotes matmul. Since we used `a->cacheAfter()` to create the smem tile, that smem may be used not only in the matmul but also in the addition in the epilogue. In that case we cannot re-use a for the epilogue smem. A conservative check that there are no other uses of `a` or `b` is currently implemented in order to guarantee re-use. A less conservative sufficient (but still not necessary) condition is that any other use of `a` or `b` is a producer of `b` or `a` respectively; this is not implemented yet.

### CODE REVIEW

**Comment by csarofeen** [comment]
The check reminds me a lot of how we check persistent buffer usage.

**Comment by naoyam** [comment]
@jacobhinkle Do we have tests?

File: csrc/scheduler/mma_utils.cpp
Comment (zasdfgbnm) [inline csrc/scheduler/mma_utils.cpp:79]: Should we add a case that, if reuse and no-reuse provides the same occupancy, then we don't promote reuse because this will save a sync?

**Review by zasdfgbnm — COMMENTED** [review]
Generally LGTM, but will leave it to @drzejan2 for approval. Also, could you run `/build/nvfuser_bench --benchmark_filter=Matmul` for reporting the perf on these benchmarks, and also work with @mmigdal-nv for a more thorough perf evaluation?

**Comment by jacobhinkle** [comment]
> Generally LGTM, but will leave it to @drzejan2 for approval. Also, could you run `/build/nvfuser_bench --benchmark_filter=Matmul` for reporting the perf on these benchmarks, and also work with @mmigdal-nv for a more thorough perf evaluation?

Will do!

File: csrc/scheduler/mma_utils.cpp
Comment (jacobhinkle) [reply csrc/scheduler/mma_utils.cpp:79]: That's a good idea. In that case I'll also need to guard `promoteReuse` with `params.use_smem_epilogue`.

Comment (jacobhinkle) [reply csrc/scheduler/mma_utils.cpp:79]: Actually, now I understand your comment. Yes we'll need a separate parameter indicating whether to reuse memory or not, in addition to `use_smem_epilogue`. I'll push something in a moment.

File: csrc/scheduler/matmul_utils.cpp
Comment (drzejan2) [inline csrc/scheduler/matmul_utils.cpp:449]: A minor suggestion:

How about passing `const mma_utils::RolesMap* roles_map` (not the optional, but the real object)?
This will allow us dropping last three parameters as they only depend on roles map. We can access required data from within `mma_utils::generateSharedMemoryEpilogueHeuristics(...)` so all smem related activities are done within this function.

**Review by drzejan2 — COMMENTED** [review]
Changes look good to me.

I will approve this MR when use case mentioned by @zasdfgbnm is supported ([link](https://github.com/NVIDIA/Fuser/pull/770/files#r1308057627)).

File: csrc/scheduler/matmul_utils.cpp
Comment (jacobhinkle) [reply csrc/scheduler/matmul_utils.cpp]: Good suggestion. I did that, but found that the tests call `generateSharedMemoryEpilogueHeuristics` directly in a way that was a bit ugly to work around since in some cases like in `test_matmul_sass.cpp` the params are used independently of a `Fusion`; in that case I'd need to either create a `Fusion` just to create some params, or move the call to `generateSharedMemoryEpilogueHeuristics` into `getSASS`. The former is a bit distracting and the latter complicates skipping the epilogue test when smem is not used. So I figured the simplest approach is to just overload the function so you can just pass a `RolesMap`, which is done in `getMatmulHeuristics`, but keep the original also which is used in the tests.

**Comment by jacobhinkle** [comment]
> I will approve this MR when use case mentioned by @zasdfgbnm is supported ([link](https://github.com/NVIDIA/Fuser/pull/770/files#r1308057627)).

Supported now. The heuristic now holds `bool promote_prologue_smem_reuse` in addition to `use_smem_epilogue`. `promote_prologue_smem_reuse` is only `true` if re-using smem would increase occupancy, since otherwise we should avoid adding a `__syncthreads()`.

File: csrc/scheduler/matmul_heuristic.h
Comment (zasdfgbnm) [inline csrc/scheduler/matmul_heuristic.h:98]: Could you also update `sameAs`?

Comment (jacobhinkle) [reply csrc/scheduler/matmul_heuristic.h:98]: Updated `sameAs` along with `hash`.

File: test/test_gpu_tensorcore.cpp
Comment (zasdfgbnm) [inline test/test_gpu_tensorcore.cpp:4001]: Is `smem_allocs.at(0)` A and `smem_allocs.at(1)` B, and `smem_allocs.at(1)` C? So B is reusing A's memory, and C is reusing B's memory?

Comment (jacobhinkle) [reply test/test_gpu_tensorcore.cpp]: Yes that's right. I improved the comment here a bit.

Comment (zasdfgbnm) [reply test/test_gpu_tensorcore.cpp]: I see. So we allocate B first, then A. This makes sense. Thanks for the explanation!

Comment (zasdfgbnm) [reply test/test_gpu_tensorcore.cpp]: Wait, is it guaranteed that which of A vs B is allocated first? If A is allocated first, then will `smem_allocs.at(1)` no longer be zero here? I think we should remove the check for `smem_allocs.at(1)`.

Comment (jacobhinkle) [reply test/test_gpu_tensorcore.cpp]: Very good question since their lifetimes end at the same time point. We break ties like that by ordering by `name()`, so `B` will be ordered after `A`, leading to this order consistently. However, that's not exactly clear so maybe we could just check that `C` is placed at `0` and that either `A` or `B` is at 0.

**Review by drzejan2 — COMMENTED** [review]
Functionally everything is sound, will approve when a bug in hashing function is fixed.

File: csrc/scheduler/matmul_heuristic.h
Comment (jacobhinkle) [reply csrc/scheduler/matmul_heuristic.h:140]: Ah ok. Makes sense. Thanks for catching that.

File: test/test_gpu_tensorcore.cpp
Comment (liqiangxl) [inline test/test_gpu_tensorcore.cpp:4013]: we may add a skip info if `promote_prologue_smem_reuse` is false. Something like:
```
      GTEST_SKIP()
          << "Test conducted without shared memory reuse since occupany is not increased.";
```

Comment (jacobhinkle) [reply test/test_gpu_tensorcore.cpp:4013]: That would skip the whole test wouldn't it? Alternatively, I could add an `else` branch here that would just check that smem is allocated as expected in case it's not re-used.

**Review by drzejan2 — APPROVED** [review]

File: test/test_gpu_tensorcore.cpp
Comment (jacobhinkle) [reply test/test_gpu_tensorcore.cpp:4046]: Just pushed a change that handles all three cases in this test: no smem epilogue, smem epilogue without reuse, and smem epilogue with reuse. Any of those could fire depending on the target device.

Comment (liqiangxl) [reply test/test_gpu_tensorcore.cpp:4035]: I was thinking the layour shouldn't influence smem usage. So these four cases will give the same result, then we can skip all if one is failed. Your change also looks good.

**Pull request merged**
